sympy/solvers/solveset.py

```diff
@@ -1117,3 +1120,514 @@ def linsolve(system, *symbols):
+    >>> a, b, c, d = symbols('a, b, c, d', real = True)
+    >>> foo =  a + b + c + d
+    >>> bar = a*b + b*c + c*d + d*a
+    >>> foo_bar = a*b*c + b*c*d + c*d*a + d*a*b
+    >>> bar_foo = a*b*c*d -1
+    >>> system = [foo, bar, foo_bar, bar_foo]
+    >>> is_zero_dimensional(system)
+    False
+    >>> pprint(nonlinsolve(system, [a, b, c, d]))
+      -1       1
+    {(---, -d, -, {d} \ {0})}
+       d       d
+
+
+    >>> x, y = symbols('x, y', real = True)
+    >>> nonlinsolve([(x+y)**2 - 4, x + y - 2], [x, y])
+    {(-y + 2, y)}
+
+    * Non linear system having non polynomial equation, if there is
+    complex solution for particular symbol and no real solution, then
+    `nonlinsolve` returns it's general solution.
+
+    Example
+    =======
+
+    >>> from sympy.core.symbol import symbols
+    >>> from sympy import sqrt, exp, sin
+    >>> from sympy.solvers.solveset import nonlinsolve
+    >>> x, y, _n = symbols('x, y, _n')
+    >>> nonlinsolve([exp(x) - sin(y), y**2 - 4], [x, y])
+    {(log(sin(2)), 2), (ImageSet(Lambda(_n, I*(2*_n*pi + pi) +
+        log(sin(2))), Integers()), -2)}
+
+    * Non linear system having all the equations polynomial, then it
+    returns both real and complex solutions.
+
+    Example
+    =======
+
+    >>> from sympy.core.symbol import symbols
+    >>> from sympy import sqrt, exp, sin
+    >>> from sympy.solvers.solveset import nonlinsolve
+    >>> x, y = symbols('x, y')
+    >>> nonlinsolve([x**2 - 2*y**2 -2, x*y - 2], [x, y])
+    {(-2, -1), (2, 1), (-sqrt(2)*I, sqrt(2)*I), (sqrt(2)*I, -sqrt(2)*I)}
+
+
+    * If system is linear positive dimensional system, `nonlinsolve` can
+    solve this system also, since it is using `groebner` method. But it
+    is not recommended to solve linear system using `nonlinsolve`, because
+    `linsolve` is better for all kind of linear system.
+
+    Examples
+    ========
+
+    >>> from sympy.core.symbol import symbols
+    >>> from sympy import pprint
+    >>> from sympy.solvers.solveset import nonlinsolve
+    >>> x, y, z = symbols('x, y, z', real = True)
+    >>> pprint(nonlinsolve([x + 2*y -z - 3, x - y - 4*z + 9 , y + z - 4], [x, y, z]))
+    {(3*z - 5, -z + 4, z)}
+
+
+    More examples:
+    =============
+
+    >>> from sympy.core.symbol import symbols
+    >>> from sympy import sqrt
+    >>> from sympy.solvers.solveset import nonlinsolve
+    >>> x, y, z = symbols('x, y, z', real = True)
+    >>> e1 = sqrt(x**2 + y**2) - 10
+    >>> e2 = sqrt(y**2 + (-x + 10)**2) - 3
+    >>> nonlinsolve((e1, e2), (x, y))
+    {(191/20, -3*sqrt(391)/20), (191/20, 3*sqrt(391)/20)}
+    >>> nonlinsolve([x**2 + 2/y - 2, x + y - 3], [x, y])
+    {(1, 2), (1 + sqrt(5), -sqrt(5) + 2), (-sqrt(5) + 1, 2 + sqrt(5))}
+    >>> nonlinsolve([x**2 + 2/y - 2, x + y - 3], [y, x])
+    {(2, 1), (2 + sqrt(5), -sqrt(5) + 1), (-sqrt(5) + 2, 1 + sqrt(5))}
+
+
+    Note :
+    =======
+
+    1. `Dict` is defined in `SymPy`. `Finiteset` can't be used with
```

Comment:
We can't do `FiniteSet({key: value})`. But we can `FiniteSet(Dict({key: value}))`.